h/services/list_groups.py

```diff
@@ -93,6 +107,21 @@ def _parse_origin(self, uri):
         origin = urlparse.SplitResult(parsed.scheme, parsed.netloc, '', '', '')
         return origin.geturl() or None
 
+    def _scoped_open_groups(self, authority, document_uri):
+        """Return scoped groups for the URI and authority"""
+        origin = self._parse_origin(document_uri)
+        if not origin:
+            return []
+
+        groups = (self._session.query(models.GroupScope, models.Group)
+                      .filter(models.Group.id == models.GroupScope.group_id)
+                      .filter(models.GroupScope.origin == origin)
+                      .filter(models.Group.authority == authority)
```

Comment:
My SQLAlchemy fu is weak; I bet some of these filters could be combined?

Reply:
You can pass multiple filter expressions as arguments to the `filter` function and they are combined using an AND operator: http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.filter ,  https://stackoverflow.com/a/44702268

However I suspect that this doesn't actually make a material difference to the generated SQL and the current code is perfectly readable to me as-is. You can check though by finding the call to `create_engine` and adding an `echo=True` argument to cause SQLAlchemy to output all the SQL statements it is executing.

Reply:
I think I vote for easily-read and maintained code over severe efficiency, so assuming there isn't a performance implication here I'd like to leave as-is, then. Thanks for the input!